modules/afmongodb/tests/test-mongodb-config.c

```diff
@@ -0,0 +1,264 @@
+
+  afmongodb_dd_set_uri(mongodb, "mongodb://localhost:1234/syslog-ng");
+  _expect_uri_in_log("uri", "localhost:1234/syslog-ng", "syslog-ng", "messages");
+
+  afmongodb_dd_set_collection(mongodb, "messages2");
+  _expect_uri_in_log("collection", "127.0.0.1:27017/syslog" SAFEOPTS, "syslog", "messages2");
+}
+
+static void
+_test_uri_error(void)
+{
+  afmongodb_dd_set_uri(mongodb, "INVALID-URI");
+  _expect_error_in_log("invalid_uri", "Error parsing MongoDB URI; uri='INVALID-URI'");
+
+  afmongodb_dd_set_uri(mongodb, "mongodb://127.0.0.1:27017/");
+  _expect_error_in_log("missing_db", "Missing DB name from MongoDB URI; uri='mongodb://127.0.0.1:27017/'");
+}
+
+#if SYSLOG_NG_ENABLE_LEGACY_MONGODB_OPTIONS
+#define UNSAFEOPTS "?w=0&safe=false&socketTimeoutMS=60000&connectTimeoutMS=60000"
+
+static void
+_test_legacy_correct(void)
+{
+  GList *servers = g_list_append(NULL, g_strdup("127.0.0.2:27018"));
+  servers = g_list_append(servers, g_strdup("localhost:1234"));
+  afmongodb_dd_set_servers(mongodb, servers);
+  _expect_uri_in_log("servers_multi", "127.0.0.2:27018,localhost:1234/syslog" SAFEOPTS,
+                     "syslog", "messages");
+
+  servers = g_list_append(NULL, g_strdup("127.0.0.2"));
+  afmongodb_dd_set_servers(mongodb, servers);
+  _expect_uri_in_log("servers_single", "127.0.0.2:27017/syslog" SAFEOPTS, "syslog", "messages");
+
+  afmongodb_dd_set_host(mongodb, "localhost");
+  _expect_uri_in_log("host", "localhost:27017/syslog" SAFEOPTS, "syslog", "messages");
+
+  afmongodb_dd_set_host(mongodb, "localhost");
+  afmongodb_dd_set_port(mongodb, 1234);
+  _expect_uri_in_log("host_port", "localhost:1234/syslog" SAFEOPTS, "syslog", "messages");
+
+  afmongodb_dd_set_port(mongodb, 27017);
+  _expect_uri_in_log("port_default", "127.0.0.1:27017/syslog" SAFEOPTS, "syslog", "messages");
+
+  afmongodb_dd_set_port(mongodb, 1234);
+  _expect_uri_in_log("port", "127.0.0.1:1234/syslog" SAFEOPTS, "syslog", "messages");
+
+  afmongodb_dd_set_path(mongodb, "/tmp/mongo.sock");
+  _expect_uri_in_log("path", "/tmp/mongo.sock" SAFEOPTS, "tmp/mongo.sock", "messages");
+
+  afmongodb_dd_set_database(mongodb, "syslog-ng");
+  _expect_uri_in_log("database", "127.0.0.1:27017/syslog-ng" SAFEOPTS, "syslog-ng", "messages");
+
+  afmongodb_dd_set_safe_mode(mongodb, TRUE);
+  _expect_uri_in_log("safe_mode_true", "127.0.0.1:27017/syslog" SAFEOPTS, "syslog", "messages");
+
+  afmongodb_dd_set_safe_mode(mongodb, FALSE);
+  _expect_uri_in_log("safe_mode_false", "127.0.0.1:27017/syslog" UNSAFEOPTS, "syslog", "messages");
+
+  afmongodb_dd_set_user(mongodb, "user");
+  afmongodb_dd_set_password(mongodb, "password");
+  _expect_uri_in_log("user_password", "user:password@127.0.0.1:27017/syslog" SAFEOPTS, "syslog", "messages");
+
+  afmongodb_dd_set_collection(mongodb, "messages2");
+  afmongodb_dd_set_safe_mode(mongodb, FALSE);
+  _expect_uri_in_log("collection_safe_mode", "127.0.0.1:27017/syslog" UNSAFEOPTS, "syslog", "messages2");
+}
```

Comment:
If the legacy stuff would have its own module we probably would not have to initialize an entire destination for tests.